test/rtpw.c

```diff
@@ -139,565 +133,568 @@ volatile int interrupted = 0;
-    } 
-    policy.ssrc.type  = ssrc_specific;
-    policy.ssrc.value = ssrc;
-    policy.key = (uint8_t *)key;
-    policy.ekt  = NULL;
-    policy.next = NULL;
-    policy.window_size = 128;
-    policy.allow_repeat_tx = 0;
-    policy.rtp.sec_serv = sec_servs;
-    policy.rtcp.sec_serv = sec_serv_none;  /* we don't do RTCP anyway */
-
-    if (gcm_on && tag_size != 8) {
-	policy.rtp.auth_tag_len = tag_size;
-    }
+            } else {
+                srtp_crypto_policy_set_null_cipher_hmac_sha1_80(&policy.rtp);
+                srtp_crypto_policy_set_rtcp_default(&policy.rtcp);
+            }
+            break;
+        default:
+            printf("error: unknown security service requested\n");
+            return -1;
+        }
+        policy.ssrc.type = ssrc_specific;
+        policy.ssrc.value = ssrc;
+        policy.key = (uint8_t *)key;
+        policy.ekt = NULL;
+        policy.next = NULL;
+        policy.window_size = 128;
+        policy.allow_repeat_tx = 0;
+        policy.rtp.sec_serv = sec_servs;
+        policy.rtcp.sec_serv = sec_serv_none; /* we don't do RTCP anyway */
+
+        if (gcm_on && tag_size != 8) {
+            policy.rtp.auth_tag_len = tag_size;
+        }
 
-    /*
-     * read key from hexadecimal or base64 on command line into an octet string
-     */
-    if (b64_input) {
-        int pad;
-        expected_len = (policy.rtp.cipher_key_len*4)/3;
-        len = base64_string_to_octet_string(key, &pad, input_key, expected_len);
-        if (pad != 0) {
-          fprintf(stderr, "error: padding in base64 unexpected\n");
-          exit(1);
+        /*
+         * read key from hexadecimal or base64 on command line into an octet
+         * string
+         */
+        if (b64_input) {
+            int pad;
+            expected_len = (policy.rtp.cipher_key_len * 4) / 3;
+            len = base64_string_to_octet_string(key, &pad, input_key,
+                                                expected_len);
+            if (pad != 0) {
+                fprintf(stderr, "error: padding in base64 unexpected\n");
+                exit(1);
+            }
+        } else {
+            expected_len = policy.rtp.cipher_key_len * 2;
+            len = hex_string_to_octet_string(key, input_key, expected_len);
+        }
+        /* check that hex string is the right length */
+        if (len < expected_len) {
+            fprintf(stderr, "error: too few digits in key/salt "
+                            "(should be %d digits, found %d)\n",
+                    expected_len, len);
+            exit(1);
+        }
+        if ((int)strlen(input_key) > policy.rtp.cipher_key_len * 2) {
+            fprintf(stderr, "error: too many digits in key/salt "
+                            "(should be %d hexadecimal digits, found %u)\n",
+                    policy.rtp.cipher_key_len * 2, (unsigned)strlen(input_key));
+            exit(1);
         }
+
+        printf("set master key/salt to %s/", octet_string_hex_string(key, 16));
```

Comment:
I can see this is old code, but I wonder why there are two printf() statements to do the work one could have done?

Reply:
Not sure there is a good answer to that. Feel free to clean up.